File: src/test/java/com/gemini/krakenbot/service/PortfolioManagerFiatCorrectionTest.java
package com.gemini.krakenbot.service;

import static org.mockito.Mockito.*;
import java.util.ArrayList;

import java.math.BigDecimal;

import com.gemini.krakenbot.model.PortfolioStats;
import com.gemini.krakenbot.service.impl.*;

import com.gemini.krakenbot.config.Allocation;
import com.gemini.krakenbot.config.AppConfig;
import com.gemini.krakenbot.repository.PortfolioStatsRepository;
import com.gemini.krakenbot.repository.impl.PortfolioStatsRepositoryImpl;
import org.junit.jupiter.api.Test;
import org.springframework.test.util.ReflectionTestUtils;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class PortfolioManagerFiatCorrectionTest {

        @Test
        void testDistributeFiatCorrection_Deposit_OnlyBuysUnderweight() {
                // Setup
                KrakenService krakenService = mock(KrakenService.class);
                ConfigService configService = mock(ConfigService.class);
                TradeHistoryService tradeHistoryService = mock(TradeHistoryService.class);

                PortfolioStatsRepository repo = mock(PortfolioStatsRepositoryImpl.class);
                when(repo.load())
                                .thenReturn(new PortfolioStats(BigDecimal.ZERO));
                PortfolioManagerImpl portfolioManager = new PortfolioManagerImpl(krakenService, configService,
                                tradeHistoryService, repo);

                // Config Mock
                Allocation allocA = new Allocation("A", 50.0);
                Allocation allocB = new Allocation("B", 50.0); // Assuming USD is implicit or handled, or strictly these
                                                               // two
                // The method iterates over config allocations.

                List<Allocation> allAllocations = List.of(allocA, allocB);
                AppConfig mockConfig = mock(AppConfig.class);
                when(mockConfig.allocations()).thenReturn(allAllocations);
                when(configService.getConfig()).thenReturn(mockConfig);

                // Inputs
                BigDecimal usdDev = BigDecimal.valueOf(100.0); // Positive = Deposit/Surplus -> Need to BUY
                Map<String, BigDecimal> allDevs = new HashMap<>();
                // Asset A is Overweight (positive dev) -> Should NOT buy
                allDevs.put("A", BigDecimal.valueOf(10.0));
                // Asset B is Underweight (negative dev) -> Should BUY
                allDevs.put("B", BigDecimal.valueOf(-10.0));

                Map<String, BigDecimal> buyOrders = new HashMap<>();
                Map<String, BigDecimal> sellOrders = new HashMap<>();

                // Invoke private method
                ReflectionTestUtils.invokeMethod(portfolioManager, "distributeFiatCorrection",
                                usdDev, allDevs, buyOrders, sellOrders, new ArrayList<String>());

                // Verify
                // Expect Buy for B
                assertTrue(buyOrders.containsKey("B"), "Should buy B because it is underweight");
                // Expect NO Buy for A
                assertEquals(0, buyOrders.getOrDefault("A", BigDecimal.ZERO).compareTo(BigDecimal.ZERO),
                                "Should NOT buy A because it is overweight");

                // Also verify sells are empty
                assertTrue(sellOrders.isEmpty());
        }

        @Test
        void testDistributeFiatCorrection_Withdrawal_OnlySellsOverweight() {
                // Setup similar to above
                KrakenService krakenService = mock(KrakenService.class);
                ConfigService configService = mock(ConfigService.class);
                TradeHistoryService tradeHistoryService = mock(TradeHistoryService.class);

                PortfolioStatsRepository repo = mock(PortfolioStatsRepositoryImpl.class);
                when(repo.load())
                                .thenReturn(new PortfolioStats(BigDecimal.ZERO));
                PortfolioManagerImpl portfolioManager = new PortfolioManagerImpl(krakenService, configService,
                                tradeHistoryService, repo);

                List<Allocation> allAllocations = List.of(new Allocation("A", 50.0), new Allocation("B", 50.0));
                AppConfig mockConfig = mock(AppConfig.class);
                when(mockConfig.allocations()).thenReturn(allAllocations);
                when(configService.getConfig()).thenReturn(mockConfig);

                // Inputs
                BigDecimal usdDev = BigDecimal.valueOf(-100.0); // Negative = Withdrawal/Shortage -> Need to SELL
                Map<String, BigDecimal> allDevs = new HashMap<>();
                // Asset A is Overweight (positive dev) -> Should SELL
                allDevs.put("A", BigDecimal.valueOf(10.0));
                // Asset B is Underweight (negative dev) -> Should NOT SELL
                allDevs.put("B", BigDecimal.valueOf(-10.0));

                Map<String, BigDecimal> buyOrders = new HashMap<>();
                Map<String, BigDecimal> sellOrders = new HashMap<>();

                // Invoke private method
                ReflectionTestUtils.invokeMethod(portfolioManager, "distributeFiatCorrection",
                                usdDev, allDevs, buyOrders, sellOrders, new ArrayList<String>());

                // Verify
                // Expect Sell for A
                assertTrue(sellOrders.containsKey("A"), "Should sell A because it is overweight");
                // Expect NO Sell for B
                assertEquals(0, sellOrders.getOrDefault("B", BigDecimal.ZERO).compareTo(BigDecimal.ZERO),
                                "Should NOT sell B because it is underweight");

                // Also verify buys are empty
                assertTrue(buyOrders.isEmpty());
        }

        @Test
        void testDistributeFiatCorrection_ProportionalDistribution() {
                // Setup
                KrakenService krakenService = mock(KrakenService.class);
                ConfigService configService = mock(ConfigService.class);
                TradeHistoryService tradeHistoryService = mock(TradeHistoryService.class);

                PortfolioStatsRepository repo = mock(PortfolioStatsRepositoryImpl.class);
                when(repo.load())
                                .thenReturn(new PortfolioStats(BigDecimal.ZERO));
                PortfolioManagerImpl portfolioManager = new PortfolioManagerImpl(krakenService, configService,
                                tradeHistoryService, repo);

                // Allocations (targets don't matter for this logic anymore, but needed for
                // config mock)
                List<Allocation> allAllocations = List.of(
                                new Allocation("A", 30.0),
                                new Allocation("B", 30.0),
                                new Allocation("C", 40.0));
                AppConfig mockConfig = mock(AppConfig.class);
                when(mockConfig.allocations()).thenReturn(allAllocations);
                when(configService.getConfig()).thenReturn(mockConfig);

                // Inputs
                BigDecimal usdDev = BigDecimal.valueOf(100.0); // $100 Surplus to spend
                Map<String, BigDecimal> allDevs = new HashMap<>();

                // Asset A: Large Deficit (-$200)
                allDevs.put("A", BigDecimal.valueOf(-200.0));
                // Asset B: Small Deficit (-$50)
                allDevs.put("B", BigDecimal.valueOf(-50.0));
                // Asset C: Surplus (+$50) -> Should be ignored
                allDevs.put("C", BigDecimal.valueOf(50.0));

                Map<String, BigDecimal> buyOrders = new HashMap<>();
                Map<String, BigDecimal> sellOrders = new HashMap<>();

                // Invoke
                ReflectionTestUtils.invokeMethod(portfolioManager, "distributeFiatCorrection",
                                usdDev, allDevs, buyOrders, sellOrders, new ArrayList<String>());

                // Verify Total Weight = 200 + 50 = 250
                // A Share = (200 / 250) * 100 = 80
                // B Share = (50 / 250) * 100 = 20

                BigDecimal boughtA = buyOrders.getOrDefault("A", BigDecimal.ZERO);
                BigDecimal boughtB = buyOrders.getOrDefault("B", BigDecimal.ZERO);

                assertEquals(0, boughtA.compareTo(BigDecimal.valueOf(80.0)),
                                "Asset A should get 80% (80.0) of surplus");
                assertEquals(0, boughtB.compareTo(BigDecimal.valueOf(20.0)),
                                "Asset B should get 20% (20.0) of surplus");

                // C should be 0
                assertEquals(0, buyOrders.getOrDefault("C", BigDecimal.ZERO).compareTo(BigDecimal.ZERO));
        }
}

File: src/test/kotlin/com/gemini/krakenbot/service/PortfolioManagerFiatCorrectionTest.kt
package com.gemini.krakenbot.service

import com.gemini.krakenbot.config.Allocation
import com.gemini.krakenbot.config.AppConfig
import com.gemini.krakenbot.config.KrakenCredentials
import com.gemini.krakenbot.config.Settings
import com.gemini.krakenbot.repository.PortfolioStatsRepository
import com.gemini.krakenbot.service.impl.PortfolioAnalyzerImpl
import io.kotest.core.spec.IsolationMode
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.booleans.shouldBeTrue
import io.kotest.matchers.comparables.shouldBeEqualComparingTo
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import java.math.BigDecimal

class PortfolioManagerFiatCorrectionTest : StringSpec() {

    override fun isolationMode() = IsolationMode.InstancePerTest

    private fun makePortfolioAnalyzer(vararg allocs: Allocation): PortfolioAnalyzer {
        val configService = mockk<ConfigService>(relaxed = true)
        val repo = mockk<PortfolioStatsRepository>(relaxed = true)

        val config = AppConfig(
            kraken = KrakenCredentials(apiKey = "k", privateKey = "s"),
            settings = Settings(
                loopDelaySeconds = 60L,
                deviationTriggerPercent = 2.0,
                dustThresholdUSD = 1.0,
                dryRun = false,
                fiatMaxDrawdown = 0.0,
                fiatDeploymentExponent = 1.0,
            ),
            allocations = allocs.toList(),
        )
        every { configService.getConfig() } returns config

        return PortfolioAnalyzerImpl(
            krakenService = FakeKrakenService(),
            configService = configService,
            portfolioStatsRepository = repo,
        )
    }

    init {
        "testDistributeFiatCorrection_Deposit_OnlyBuysUnderweight" {
            val portfolioAnalyzer = makePortfolioAnalyzer(
                Allocation("A", 50.0),
                Allocation("B", 50.0),
            )

            val usdDev = BigDecimal.valueOf(100.0)
            // A is overweight (+10), B is underweight (-10) → only B should receive a buy
            val allDevs = mapOf(
                "A" to BigDecimal.valueOf(10.0),
                "B" to BigDecimal.valueOf(-10.0),
            )
            val buyOrders = mutableMapOf<String, BigDecimal>()
            val sellOrders = mutableMapOf<String, BigDecimal>()

            portfolioAnalyzer.distributeFiatCorrection(
                usdDev = usdDev,
                allDevs = allDevs,
                buyOrders = buyOrders,
                sellOrders = sellOrders,
                actionLog = mutableListOf(),
            )

            buyOrders.containsKey("B").shouldBeTrue()
            buyOrders.getOrDefault("A", BigDecimal.ZERO).shouldBeEqualComparingTo(BigDecimal.ZERO)
            sellOrders.isEmpty().shouldBeTrue()
        }

        "testDistributeFiatCorrection_Withdrawal_OnlySellsOverweight" {
            val portfolioAnalyzer = makePortfolioAnalyzer(
                Allocation("A", 50.0),
                Allocation("B", 50.0),
            )

            val usdDev = BigDecimal.valueOf(-100.0)
            // A is overweight (+10), B is underweight (-10) → only A should receive a sell
            val allDevs = mapOf(
                "A" to BigDecimal.valueOf(10.0),
                "B" to BigDecimal.valueOf(-10.0),
            )
            val buyOrders = mutableMapOf<String, BigDecimal>()
            val sellOrders = mutableMapOf<String, BigDecimal>()

            portfolioAnalyzer.distributeFiatCorrection(
                usdDev = usdDev,
                allDevs = allDevs,
                buyOrders = buyOrders,
                sellOrders = sellOrders,
                actionLog = mutableListOf(),
            )

            sellOrders.containsKey("A").shouldBeTrue()
            sellOrders.getOrDefault("B", BigDecimal.ZERO).shouldBeEqualComparingTo(BigDecimal.ZERO)
            buyOrders.isEmpty().shouldBeTrue()
        }

        "testDistributeFiatCorrection_ProportionalDistribution" {
            val portfolioAnalyzer = makePortfolioAnalyzer(
                Allocation("A", 30.0),
                Allocation("B", 30.0),
                Allocation("C", 40.0),
            )

            val usdDev = BigDecimal.valueOf(100.0)
            // A is underweight by 200, B by 50; C is overweight by 50.
            // With a 100 USD deposit, distribution should be proportional to underweight magnitude:
            //   A gets 200/(200+50) * 100 = 80, B gets 50/(200+50) * 100 = 20.
            val allDevs = mapOf(
                "A" to BigDecimal.valueOf(-200.0),
                "B" to BigDecimal.valueOf(-50.0),
                "C" to BigDecimal.valueOf(50.0),
            )
            val buyOrders = mutableMapOf<String, BigDecimal>()
            val sellOrders = mutableMapOf<String, BigDecimal>()

            portfolioAnalyzer.distributeFiatCorrection(
                usdDev = usdDev,
                allDevs = allDevs,
                buyOrders = buyOrders,
                sellOrders = sellOrders,
                actionLog = mutableListOf(),
            )

            buyOrders.getOrDefault("A", BigDecimal.ZERO).shouldBeEqualComparingTo(BigDecimal.valueOf(80.0))
            buyOrders.getOrDefault("B", BigDecimal.ZERO).shouldBeEqualComparingTo(BigDecimal.valueOf(20.0))
            buyOrders.getOrDefault("C", BigDecimal.ZERO).shouldBeEqualComparingTo(BigDecimal.ZERO)
        }

        "testDistributeFiatCorrection_ShareUsesUsdScale" {
            val portfolioAnalyzer = makePortfolioAnalyzer(
                Allocation("A", 50.0),
                Allocation("B", 50.0),
            )

            // Uneven counter-devs produce a many-decimal ratio; share must round to USD scale 2.
            val usdDev = BigDecimal("100.00")
            val allDevs = mapOf(
                "A" to BigDecimal("-70.00"),
                "B" to BigDecimal("-30.00"),
            )
            val buyOrders = mutableMapOf<String, BigDecimal>()

            portfolioAnalyzer.distributeFiatCorrection(
                usdDev = usdDev,
                allDevs = allDevs,
                buyOrders = buyOrders,
                sellOrders = mutableMapOf(),
                actionLog = mutableListOf(),
            )

            buyOrders.getValue("A").scale() shouldBe 2
            buyOrders.getValue("B").scale() shouldBe 2
            buyOrders.getValue("A").shouldBeEqualComparingTo(BigDecimal("70.00"))
            buyOrders.getValue("B").shouldBeEqualComparingTo(BigDecimal("30.00"))
        }

        // CQ-3-26 / #76: a $0.00 rounded share must never be enqueued as an order.
        "testDistributeFiatCorrection_ZeroRoundedShareNotEnqueued_Deposit" {
            val portfolioAnalyzer = makePortfolioAnalyzer(
                Allocation("TINY", 1.0),
                Allocation("BIG", 99.0),
            )

            // TINY ratio = 0.001/100 → share = $1.00 * 0.00001 = $0.00001 → $0.00 at USD scale.
            val usdDev = BigDecimal("1.00")
            val allDevs = mapOf(
                "TINY" to BigDecimal("-0.001"),
                "BIG" to BigDecimal("-99.999"),
            )
            val buyOrders = mutableMapOf<String, BigDecimal>()
            val sellOrders = mutableMapOf<String, BigDecimal>()

            portfolioAnalyzer.distributeFiatCorrection(
                usdDev = usdDev,
                allDevs = allDevs,
                buyOrders = buyOrders,
                sellOrders = sellOrders,
                actionLog = mutableListOf(),
            )

            buyOrders.containsKey("TINY") shouldBe false
            buyOrders.values.none { it.signum() == 0 }.shouldBeTrue()
            sellOrders.isEmpty().shouldBeTrue()
            (
                buyOrders.values.fold(BigDecimal.ZERO, BigDecimal::add)
                    .compareTo(usdDev.abs()) <= 0
                ).shouldBeTrue()
        }

        // CQ-3-26 / #76: same zero-share filter on withdrawal sells.
        "testDistributeFiatCorrection_ZeroRoundedShareNotEnqueued_Withdrawal" {
            val portfolioAnalyzer = makePortfolioAnalyzer(
                Allocation("TINY", 1.0),
                Allocation("BIG", 99.0),
            )

            val usdDev = BigDecimal("-1.00")
            val allDevs = mapOf(
                "TINY" to BigDecimal("0.001"),
                "BIG" to BigDecimal("99.999"),
            )
            val buyOrders = mutableMapOf<String, BigDecimal>()
            val sellOrders = mutableMapOf<String, BigDecimal>()

            portfolioAnalyzer.distributeFiatCorrection(
                usdDev = usdDev,
                allDevs = allDevs,
                buyOrders = buyOrders,
                sellOrders = sellOrders,
                actionLog = mutableListOf(),
            )

            sellOrders.containsKey("TINY") shouldBe false
            sellOrders.values.none { it.signum() == 0 }.shouldBeTrue()
            buyOrders.isEmpty().shouldBeTrue()
            (
                sellOrders.values.fold(BigDecimal.ZERO, BigDecimal::add)
                    .compareTo(usdDev.abs()) <= 0
                ).shouldBeTrue()
        }

        // CQ-3-26 / #76: HALF_UP USD rounding must not let shares sum exceed |usdDev|.
        "testDistributeFiatCorrection_RoundedSharesDoNotExceedUsdDev" {
            val portfolioAnalyzer = makePortfolioAnalyzer(
                Allocation("A", 50.0),
                Allocation("B", 50.0),
            )

            // Equal ratios: each share = $0.05 * 0.5 = $0.025 → $0.03 after USD scale;
            // naive enqueue would sum to $0.06 > $0.05.
            val usdDev = BigDecimal("0.05")
            val allDevs = mapOf(
                "A" to BigDecimal("-1.00"),
                "B" to BigDecimal("-1.00"),
            )
            val buyOrders = mutableMapOf<String, BigDecimal>()

            portfolioAnalyzer.distributeFiatCorrection(
                usdDev = usdDev,
                allDevs = allDevs,
                buyOrders = buyOrders,
                sellOrders = mutableMapOf(),
                actionLog = mutableListOf(),
            )

            buyOrders.values.none { it.signum() == 0 }.shouldBeTrue()
            (
                buyOrders.values.fold(BigDecimal.ZERO, BigDecimal::add)
                    .compareTo(usdDev.abs()) <= 0
                ).shouldBeTrue()
        }
    }
}
